spring-integration-core/src/main/java/org/springframework/integration/support/management/graph/EndpointNode.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.support.management.graph;
+
+/**
+ * Base class for all endpoints.
+ *
+ * @author Gary Russell
+ * @since 4.3
+ *
+ */
+public abstract class EndpointNode extends IntegrationNode {
+
+	private final String output;
+
+	private final String input;
```

Comment:
I think we still is good in providing one more intermediate hierarchy member - `ProducerEndpointNode` - just with the `output`. And the `input` property would be only the part of `MessageHandlerNode`.
It's a bit confusing to see an `input` on the `MessageSource` or `MessageProducer`...